runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/DistributionData.java

```diff
@@ -29,36 +42,225 @@
+          .build();
+  // k = 256 should yield an approximate error ε of less than 1%
+  private static final int SKETCH_SUMMARY_SIZE = 256;
 
-  public abstract long sum();
+  private long sum;
+  private long count;
+  private long min;
+  private long max;
+  private Map<Double, Double> percentiles;
+  private transient UpdateDoublesSketch sketch;
 
-  public abstract long count();
+  /** Backward compatible static factory method. */
+  public static DistributionData empty() {
+    return new DistributionData();
+  }
 
-  public abstract long min();
+  /** Static factory method primary used for testing. */
+  @VisibleForTesting
+  public static DistributionData create(long sum, long count, long min, long max) {
+    return create(sum, count, min, max, ImmutableMap.of());
+  }
 
-  public abstract long max();
+  /** Static factory method primary used for testing. */
+  @VisibleForTesting
+  public static DistributionData create(
+      long sum, long count, long min, long max, Map<Double, Double> percentiles) {
+    return new DistributionData(sum, count, min, max, percentiles);
+  }
 
-  public static final DistributionData EMPTY = create(0, 0, Long.MAX_VALUE, Long.MIN_VALUE);
+  private DistributionData() {
+    this(0L, 0L, Long.MAX_VALUE, Long.MIN_VALUE, ImmutableMap.of());
+    final DoublesSketchBuilder doublesSketchBuilder = new DoublesSketchBuilder();
+    this.sketch = doublesSketchBuilder.setK(SKETCH_SUMMARY_SIZE).build();
+  }
 
-  public static DistributionData create(long sum, long count, long min, long max) {
-    return new AutoValue_DistributionData(sum, count, min, max);
+  private DistributionData(
+      long sum, long count, long min, long max, Map<Double, Double> percentiles) {
+    this.sketch = null;
+    this.sum = sum;
+    this.count = count;
+    this.min = min;
+    this.max = max;
+    this.percentiles = percentiles;
   }
 
   public static DistributionData singleton(long value) {
-    return create(value, 1, value, value);
+    final DistributionData distributionData = new DistributionData();
+    distributionData.update(value);
+    return distributionData;
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////////////////////////
+  // Getters
+
+  public long sum() {
+    return sum;
+  }
+
+  public long count() {
+    return count;
+  }
+
+  public long min() {
+    return min;
+  }
+
+  public long max() {
+    return max;
+  }
+
+  public Map<Double, Double> percentiles() {
+    return percentiles;
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Updates the distribution with a value.
+   *
+   * @param value value to update the distribution with.
+   */
+  public void update(long value) {
+    ++count;
+    min = Math.min(min, value);
+    max = Math.max(max, value);
+    sum += value;
+    if (sketch != null) {
+      sketch.update(value);
+    }
+    percentiles = computePercentiles();
```

Comment:
Do we need to compute the percentiles for every update? Not sure whether we only need to do it during extractResult().